File: python/agent_disclosure/emit.py
"""Native Python emitter. Port of the signing half of `src/attestation.ts`.

Mints/loads an ed25519 signing identity and produces signed envelopes whose
bytes are identical to the TS reference: the value is the hex ed25519 signature
over the UTF-8 bytes of `canonicalize(disclosure)`, and the agentId / envelope
publicKey is the raw 32-byte public key as hex.

Key loading mirrors `agentKeyFromPrivateHex`: the fixture stores the private key
as PKCS8 DER hex. For ed25519 the DER is the fixed prefix
`302e020100300506032b657004220420` followed by the 32-byte seed, so the seed is
the trailing 32 bytes and `Ed25519PrivateKey.from_private_bytes` reconstructs it.
"""

from cryptography.hazmat.primitives.asymmetric import ed25519
from cryptography.hazmat.primitives import serialization

from .canonical import canonicalize


class AgentKey:
    """An ed25519 signing identity: the private key plus its raw public hex
    (the agentId + envelope publicKey)."""

    def __init__(self, private_key: ed25519.Ed25519PrivateKey):
        self.private_key = private_key
        self.public_key_hex = private_key.public_key().public_bytes(
            encoding=serialization.Encoding.Raw,
            format=serialization.PublicFormat.Raw,
        ).hex()


def generate_agent_key() -> AgentKey:
    """Mint a fresh signing identity."""
    return AgentKey(ed25519.Ed25519PrivateKey.generate())


def agent_key_from_private_hex(hex_str: str) -> AgentKey:
    """Reconstruct a signing identity from a persisted PKCS8 DER private key (hex).

    The 32-byte seed is the trailing 32 bytes of the DER (after the
    `302e020100300506032b657004220420` ed25519 PKCS8 prefix)."""
    der = bytes.fromhex(hex_str)
    seed = der[-32:]
    return AgentKey(ed25519.Ed25519PrivateKey.from_private_bytes(seed))


def sign_message(message: str, priv: AgentKey) -> str:
    """Hex ed25519 signature over the UTF-8 bytes of `message`."""
    return priv.private_key.sign(message.encode("utf-8")).hex()


def sign_disclosure(disclosure: dict, priv: AgentKey) -> dict:
    """Sign a disclosure, returning the signed envelope. agentId and the envelope
    publicKey are the derived raw public hex; the value signs over
    `canonicalize(disclosure)` (the disclosure is signed as-is — set its agentId
    to the key's public hex before signing to satisfy the identity binding)."""
    return {
        "disclosure": disclosure,
        "signature": {
            "algorithm": "ed25519",
            "publicKey": priv.public_key_hex,
            "value": sign_message(canonicalize(disclosure), priv),
        },
    }

File: go/emit.go
package agentdisclosure

import (
	"crypto/ed25519"
	"encoding/hex"
)

// AgentKeyPair is a Go-side agent signing identity. PublicKeyHex is the raw
// 32-byte ed25519 public key as hex - this is both the agentId and the envelope
// signature.publicKey (SPEC.md section 5). Mirrors the TS AgentKeyPair.
type AgentKeyPair struct {
	Private      ed25519.PrivateKey
	Public       ed25519.PublicKey
	PublicKeyHex string
}

// pkcs8Ed25519Prefix is the PKCS8 DER prefix that precedes the 32-byte seed in a
// stored ed25519 private key (the form the TS reference exports). The full DER is
// this prefix + the 32-byte seed; ed25519.NewKeyFromSeed reconstructs the pair.
const pkcs8Ed25519Prefix = "302e020100300506032b657004220420"

// GenerateAgentKeyPair mints a fresh agent signing identity.
func GenerateAgentKeyPair() (AgentKeyPair, error) {
	pub, priv, err := ed25519.GenerateKey(nil)
	if err != nil {
		return AgentKeyPair{}, err
	}
	return AgentKeyPair{Private: priv, Public: pub, PublicKeyHex: hex.EncodeToString(pub)}, nil
}

// AgentKeyFromPrivateHex reconstructs a key pair from a persisted private key in
// PKCS8 DER hex (the `key.privateKeyHex` form in the conformance fixture). The
// trailing 32 bytes of the DER are the ed25519 seed.
func AgentKeyFromPrivateHex(privateKeyHex string) (AgentKeyPair, error) {
	der, err := hex.DecodeString(privateKeyHex)
	if err != nil {
		return AgentKeyPair{}, err
	}
	seed := der[len(der)-ed25519.SeedSize:]
	priv := ed25519.NewKeyFromSeed(seed)
	pub := priv.Public().(ed25519.PublicKey)
	return AgentKeyPair{Private: priv, Public: pub, PublicKeyHex: hex.EncodeToString(pub)}, nil
}

// SignMessage signs the UTF-8 bytes of message with the agent private key and
// returns the hex signature - the generic primitive the disclosure signing and
// the challenge handshake both build on (mirrors TS signMessage).
func SignMessage(message string, priv ed25519.PrivateKey) string {
	return hex.EncodeToString(ed25519.Sign(priv, []byte(message)))
}

// SignDisclosure signs a disclosure with an agent key, returning the signed
// envelope. It sets the disclosure's agentId and the signature.publicKey to the
// key's derived public hex, and signs over Canonicalize(disclosure) - byte-for-byte
// the TS signDisclosure scheme. The input disclosure map is mutated to bind agentId.
func SignDisclosure(disclosure Disclosure, key AgentKeyPair) SignedDisclosure {
	disclosure["agentId"] = key.PublicKeyHex
	return SignedDisclosure{
		Disclosure: disclosure,
		Signature: Signature{
			Algorithm: "ed25519",
			PublicKey: key.PublicKeyHex,
			Value:     SignMessage(Canonicalize(disclosure), key.Private),
		},
	}
}
